### ISSUE TITLE
Remove FZAPI

### ISSUE THREAD

**Issue by SunBlack** [PointCloudLibrary/pcl#2701]
As far as I can see FZAPI is from fotonic, but their sites seem to be down (checked it already a few days ago). We have currently just a single demo which is using this API and this demo is not even build on azure build server. As I didn't found any other for this library, maybe we could remove this demo/dependency.

**Comment by SergioRAgostinho** [comment]
Is it this one? http://fotonic.kunddemo.se/

I would be ok with this, but this is not something we can simply remove. We need to add the warning that we're planning to drop support and let the community react to it. I really have no idea how widespread the use of these cameras is.

**Comment by taketwo** [comment]
I also have no idea how popular this grabber is. What would we gain by removing it though?

**Comment by SunBlack** [comment]
> Is it this one? http://fotonic.kunddemo.se/

Ah well, didn't found this site. Just found http://www.fotonic.com. Now I found they were [acquired by Autoliv](https://cleantechnica.com/2017/09/30/autoliv-acquiring-lidar-assets-swedens-fotonic-norden-ab/).

Gain by removing: Because build server do not built this target, we don't know when it breaks. So maybe we have already not working code there. And because I don't find source of this library, we can't change it. If you find this code, we can keep it. If not I vote for removing it.

**Comment by SergioRAgostinho** [comment]
This topic made think about another one which is revising a which algorithms are currently implemented in PCL and keeping only widely accepted ones. A topic for another thread on its own. This came after the realization that last year I spent time looking at PRs from new handcrafted features which looking back in retrospective are not really that popular and are somewhat bloating the library. (Apologies for derailing the issue)

> What would we gain by removing it though?

I don't have much to add. It's a maintainability burden on our side, which to some extent is not real because if something breaks we don't have a way to fix it since none of us has the hardware :/ It would be really nice to know which algorithms/modules are most used by the community.

**Comment by taketwo** [comment]
OpenCV's answer to this is "contrib" repository. They have pretty low standards for accepting code there, and thus a lot of stuff. Occasionally they promote algorithms to the "main" repository.

**Comment by SergioRAgostinho** [comment]
I would be open for giving it a try, inclusive with the low standards. 

Alternatively we try something along the lines of the homebrew tap system. We define a certain folder structure for user extensions and provide scripts to download these upon requests into a known location (`extensions` sub folder). This will require some extra cmake magic as I can see extensions interested in defining extra targets inside our currently existing modules, some even defining their own. Each creator is responsible for maintaining it's own "tap" and keeping it up to date with the main repo. That way we focus on maintaining core features and the contributors in maintaining their contributions. Everyone is happy.

**Comment by taketwo** [comment]
Sounds good. But this would be pretty low on my priority list...

**Comment by SergioRAgostinho** [comment]
Indeed. c++14 and library wide breaking bugs is the only thing we should be talking right now. 
Basically always to go with the most desired features without contracting too much technical debt (e.g. continuing the CMake cleanup).

Popularity wise Python bindings probably comes immediately after C++14.

**Comment by SunBlack** [comment]
Another reason to remove it if we cannot build it at Azure pipelines: #2746

**Comment by taketwo** [comment]
Sorry, I still don't get why we have to remove this grabber. Yes, it is not built on CI servers so we are not entirely sure if it's still working. But that's okay. If we accidentally break it and nobody notices - fine. If someone is affected, this will be reported and (hopefully) a patch will be submitted - also fine. So no extra effort is spent maintaining it. Can you explain why #2746 requires this grabber to be removed?

If we remove the grabber, there is a chance that people will come asking: "This was a perfectly functional grabber and I used it. Why the hell did you remove it?" I don't know if this will happen, but if it does, I won't have any good answer for them.

Further, I actually see circumstantial evidence that these cameras are still in use. Check e.g. [this](https://github.com/BoomFan/fotonic_wrapper) and [this](https://github.com/RobotnikAutomation/fotonic_3dcamera) repositories, last commits in 2017 and 2016 respectively.

**Comment by SunBlack** [comment]
This is like dead code: Mostly you don't care about it. But in case you are refactoring things (like we are currently doing. e.g. by partially replacing boost by std or CMake refactoring), you will always see this code and think short about it, before you see, it is dead code (e.g. if i search for a keyword with a general text search). So you spent even for dead code a bit time. I don't know how you think about dead code in general: But we have Git, so we always take a look into old removed code and in case someone want to use it, he can restore it, refactor it and integrate it back.

Further point: PCL just support a few formats. E.g. it doesn't support LAS and E57 (both are common point cloud formats - more than e.g. even ply). So a lot of projects have their own I/O classes. With this I don't see a reason why we should keep an unmaintained I/O class, if it is still possible to load point clouds into PCL from your own code. So if someone still use this code: He can integrate this reader into his own framework and maintain it there.

> Further, I actually see circumstantial evidence that these cameras are still in use. Check e.g. [this](https://github.com/BoomFan/fotonic_wrapper) and [this](https://github.com/RobotnikAutomation/fotonic_3dcamera) repositories, last commits in 2017 and 2016 respectively.

Both has last commits before Fotonic was acquired by Autoliv. Via Google you still find API documentation, bot only on old sute and without Google Cache. So it future of this format is unknown by us.

**Comment by SunBlack** [comment]
Forgot to answer your question @taketwo

> Can you explain why #2746 requires this grabber to be removed?

In case someone is enabling warning as errors (e.g. before submitting sth.) he needs always to disable everything with FZ-API, because we cannot fix compiler warnings there as long we don't see them. So whoever is still using this will have pain in the ass with this unmaintained code. As already mentioned in my last comment: if we remove this code, it is still archived due to Git.

**Comment by taketwo** [comment]
I understand all the inconveniences caused by this. Still, even if we decide to remove the grabber, we have procedures.  To remove an API we first mark it as deprecated and keep around for some time so that the users have time to adapt. If we follow your proposal, we'll have to depart from this rule too. @SergioRAgostinho what's your opinion? @jspricke as a seasoned maintainer, what would you suggest?

**Comment by jspricke** [comment]
I think a contrib (or staging as in the Linux kernel) tree makes sense. Also it's quite common in the kernel to break a driver to make people aware that something is unmaintained and will be removed. So if no one complained after some releases, you can remove it as well. For this one I would vote for breaking it (maybe it would be enough to add a error pragma..).

**Comment by taketwo** [comment]
@jspricke thanks for the input. If you are not against this, then let it be so!

**Comment by SergioRAgostinho** [comment]
Sorry for the late input. 

On a personal side, I really don't mind having this code out of the main tree. The small short story about dead code from @SunBlack does resonate with me, with a lot more than this grabber. 

The suggestion @jspricke mentioned is indeed interesting and I wasn't expecting it to exist such a practice in the linux kernel project. It makes me feel less guilty for wanting this code gone, so I'm ok in going in that direction.

As a closing statement, I have the impression we all agree that the best solution would be the separate contrib tree, but highly unrealistic given everything else we need to do. I'm glad though that we start considering that as something nice to have.

**Comment by taketwo** [comment]
Closed by #2747.
